### ISSUE TITLE
The `setErrors` method does not set errors

### ISSUE THREAD

**Issue by sonukapoor** [angular/angular#38191]
When you set the errors of a `FormControl` (using the `setErrors` method) inside a `FormGroup`, where the `FormControl` contains validators then the errors are not set. 

```typescript
formGroup: FormGroup;

constructor(private formBuilder: FormBuilder) {
  this.formGroup = this.formBuilder.group({
    login: ["", [Validators.required]]
  });
}

ngOnInit() {
  const formLoginControl = this.formGroup.controls.login;
  formLoginControl.setErrors({ incorrect: true });
}
```

Here is a [stackblitz ](https://stackblitz.com/edit/angular-forms-set-errors-gh-issue-38191-cb9m14?file=app/app.component.ts)of the issue.

This is happening because the validation has not run yet. Since `setErrors` is called before the validation has run, the errors get overwritten by the validator. Note that this issue occurs, when your `FormControl` is part of a `FormGroup` - it does not happen on stand-alone controls that are not part of a `FormGroup`.

To overcome this, `setErrors` needs to be called in a `setTimeout`. 

Action items:

- Update the docs to make this issue clear
- Fix the underlying issue

**Comment by MohamedSahir** [comment]
@sonukapoor  regarding 38017 issues,  thanks for your solution and need to add setTimeOut for each control, I have 40 controls in a form .is there any other  recommended way for this  for time being.
![image](https://user-images.githubusercontent.com/20682302/88300400-10576800-cd21-11ea-97f5-62392f77599f.png)

**Comment by sonukapoor** [comment]
@MohamedSahir Try to use a generic approach. From the screenshot and the two examples in your demo, it seems that the properties are the same. Just the control names are different. Maybe you can create a function that would take the control name as an input: Something like:

```typescript
setFormControl(controlName: string, response: any) {
  // Set errors here
  setTimeout(() => {
    form.controls[controlName].markAsDirty();
    form.controls[controlName].markAsTouched();
    // ...
  }, 1);
  
  // Patch value here
  form.patchValue({
      [controlName]: response...,
      emitEvent: false,
      onlySelf: true
    });
}
```

**Comment by MohamedSahir** [comment]
@sonukapoor  in the stackblitz with one field it is working fine. But when comes to local implementation, the setTimeout is triggering after return the forms from service to component.

The flow of execution changed in realtime i.
1) first it executes the outside of set timeout context and return the forms . after that it executes setimeout. so the error is not because it returns the forms in the component.

**Comment by aakash14goplani** [comment]
Hello @sonukapoor 
I was also facing similar kind of issue and setTimeout did the trick! Thanks for that. I need you help in understanding how to debug this scenario ` Since setErrors is called before the validation has run, the errors get overwritten by the validator.` on the browser? How would I test setError is executing before validation has begun and how to test that errors are getting overwritten?

**Comment by Express85** [comment]
We were facing the same issue.
Trying to manipulate setErrors but still having problems.
The setTimeout was the trick.

Thanks

**Comment by durveshparmar** [comment]
@all 

I was also facing the same issue but I think it should be resolved as per following: 
```

const formLoginControl = this.formGroup.controls.login;
formLoginControl.valueChanges.subscribe(value=> {
//Condition based on value
       formLoginControl.setErrors({ incorrect: true });
})

```

It works for me properly with expected output whenever value is updated. You can add this code in ngOnInit() as per your need. I don't recommend setTimeout Solution.

**Comment by abhiabhisheik646** [comment]
for (const key of Object.keys(this.addAgent.controls)) {
      if (this.addAgent.controls[key].invalid) {
        this.addAgent.controls['email'].setErrors({'incorrect': true});
        const invalidControl = this.el.nativeElement.querySelector('[formcontrolname="' + key + '"]');
        invalidControl.focus();
        invalidControl.scrollIntoView({block: "end", behavior: "smooth"})
        break;
     }

**Comment by AnejVovcak** [comment]
Has there been any update on this?

**Comment by montella1507** [comment]
So we have 3 years anniversary here! Atleast, we have new logo.

**Comment by driabov-nix** [comment]
I think a better way of doing this is like that:
`
ngAfterViewInit(): void {
   control.setErrors({ error: true });
    control.markAsTouched();
    this.cdr.detectChanges();
  }
`

-  It sets an error after the component have run initial validation.
-  We need a `detectChanges` bc we change rendered stuff after view init.
-  This way you only trigger change detection from your current component instead of triggering it from the root of the app via `setTimeout` which is better for performance.

**Comment by herickwilke** [comment]
Experiencing the same today

**Comment by wkrell3** [comment]
So, there is a "solution" to this. It is not that the code functions how we think it should, but rather something that involves using Angular's default Validators in a different way.

Consider something like the following where you are building or digesting a form Model on the front end.

`form.forEach((formField: any) => {
        let validatorsArray = [];
        formField.validators?.forEach((validator: any) => {
          switch(validator.type){
            case 'required':
              validatorsArray.push(Validators.required);
              validatorsArray.push(this.required());
              break;
            case 'pattern':
              validatorsArray.push(this.testValue(validator));
              break;
            case 'minLength':
            case 'maxLength':
              validatorsArray.push(this.minMax(validator));
              break;
            case 'mustMatch':
              validatorsArray.push(this.customValidator(formField.value, validator.name, formField.options));
              break;
          }
        });
    Object.assign(formObj, {[formField.name]: new FormControl({value: '', disabled: false}, validatorsArray)});
});`

If you're trying to use Angular's validator, like required, you essentially have to add it (like you would normally expect) AND do something to actually correctly handle the error, like the this.required() function, which I either got from a previous post here, or from stack overflow. 

`
private required(): ValidatorFn {
    return (control: AbstractControl) => {
      let error = Validators.required(control);
      return error ? {required: 'This field is required'} : null;
    }
  }
`

Now in your template, when you're iterating through errors (and possibly assigning new errors from the server dynamically) you're good to go.

The same logic would apply to any "out of the box" validator from Angular like min, max, etc.

Essentially, add the validator to the Validators array (like Angular wants you to) AND add an additional validator that will actually fire the error accordingly.

`
<mat-error id="{{configuration.label}} error">

          <ng-container *ngFor="let error of control.errors | keyvalue">
            {{error.value}}
          </ng-container>
          
        </mat-error>
`
I came across this problem when assigning something from the server like `this.control.setErrors({alreadyExists: 'This column name already exists'})` 

Then if I removed everything from the input (trying to kick off required validator) the field would highlight red, but no text would show.

If I removed the Validators.required from the validator array, then I would lose all functionality of getting the required asterisk on the respective form field, etc.

Long story short - It's extra work (and code) but if you want to use the built in Validators AND actually use setErrors like you assume it should work, then make an additional validator for each out of the box Angular validator to actually handle showing the error.

I hope this is helpful. Thank God for everyone who actually contributed to formulating this answer, because I was stumped for a good while on it.

**Comment by JeanMeche** [comment]
I wouldn't recommend using `setErrors` to set errors. 

More than a bug I'd say it's a doc issue. It should be explicit that manually set errors are overwritten by any validation. 

So the safest way to set errors is to do it via a validator.

**Comment by ilyakonrad** [comment]
> I wouldn't recommend using `setErrors` to set errors.
> 
> More than a bug I'd say it's a doc issue. It should be explicit that manually set errors are overwritten by any validation.

I don't have any validators on my form control and I have the same issue.

I have a valid reason to use `setErrors` in a specific situation. It's much easier than creating a redundant custom validator and then having to play around with adding and removing it and also calling `updateValueAndValidity()` afterwards.

**Comment by JeanMeche** [comment]
#64063 Will add a note to the docs. 

TLDR: This is working as intended. Manually set errors get overwritten by validation.
